File: C++/ReverseLinkedList.cpp
/***********************************************************************************************
Problem      Reverse Linked List
Developer    Ian Hays
Date         09/07/2021
URL          https://leetcode.com/problems/reverse-linked-list/
Space        O(1) 
Time         O(N)
Description  We keep track of the previous node, and the next node reversing as we iterate to 
             the last node in the linked list and return.
************************************************************************************************/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(!head) return nullptr;
        ListNode* prev = nullptr;
        
        while(head){
            ListNode* next = head->next;
            head->next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }
};

File: Java/ReverseLinkedList.java
/***********************************************************************************************
Problem      Reverse Linked List
Developer    Ian Hays
Date         07/27/22
URL          https://leetcode.com/problems/reverse-linked-list/
Space        O(1) 
Time         O(N)
Description  Using a previous pointer build the result by pointing to nodes in descending order
************************************************************************************************/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode ans = null;
        ListNode prev = head;
        while(head != null){
            head = head.next;
            prev.next = ans;
            ans = prev;
            prev = head;
        }
        return ans;
    }
}
